File: swap_pair/swap_pair.cpp
#include <iostream>
using namespace std;

class LinkedList {
	private:
		struct Node {
			int val;
			Node * next = NULL;
		};
		Node* head = NULL;

	public:
		void add (int val) {
			Node* new_node = new Node();
			new_node->val = val;
			new_node->next = head;
			head = new_node;
		}

		void swap_pairs() {
			Node* node = head;
			while (node != NULL && node->next != NULL && node->next->next != NULL) {
				if (node == head) {
					Node* temp;
					temp = node->next;
					node->next = node->next->next;
					temp->next = node;
					node = temp;
					temp = temp->next;
					head = node;
					node = temp;
				}
				else {
					Node* temp1;
					temp1 = node->next->next;
					node->next->next = node->next->next->next;
					temp1->next = node->next;
					node->next = temp1;
					node = node->next->next;
				}
			}
		}
		
		void print_list() {
			Node* node = head;
			while (node != NULL) {
				cout << node->val << " ";
				node = node->next;
			}
			cout << endl;
		}
};

int main() {
	LinkedList node;
	node.add(5);
	node.add(10);
	node.add(7);
	node.add(8);
	
	node.print_list();
	node.swap_pairs();
	node.print_list();

	return 0;
}

File: swap_pair/swap_pair.py
class Node:
    def __init__(self, val, nexti=None):
        self.val = val
        self.next = nexti

    def __str__(self):
        node = self
        s = ""
        sep = ""
        while node != None:
            s += sep + str(node.val)
            node = node.next
            sep = " "
        return s

def swap_pairs(n):
    node = n
    while node != None and node.next != None and node.next.next != None:
        if node == n:
            temp = node.next
            node.next = node.next.next
            temp.next = node
            n = temp
        else:
            temp = node.next.next
            node.next.next = node.next.next.next
            temp.next = node.next
            node.next = temp
            node = node.next.next
    return n

def main():
    node = Node(6)
    node.next = Node(5)
    node.next.next = Node(4)
    node.next.next.next = Node(3)

    node = None
    for v in 6,5,4,3:
        node = Node(v,node)
    
    print('Before:',node)
    print('After :',swap_pairs(node))

    node = Node(6)
    node.next = Node(5)
    node.next.next = Node(4)

    print('Before:',node)
    print('After :',swap_pairs(node))

main()
